### ISSUE TITLE
Array index expressions in contract

### ISSUE THREAD

**Issue by Drevanoorschot** [utwente-fmt/vercors#377]
Although
```
requires (\forall  int i; i >= (partitionSize * tid) && i < (partitionSize * (tid + 1)); input[i] >= 0 && input[i] < radix);
```
and 
```
requires (\forall  int i; i >= 0 && i < partitionSize; input[(tid * partitionSize) + i] >= 0 && input[(tid * partitionSize) + i] < radix);
```
are equivalent logic-wise, the first will verify while the second will not. This particular expression is used in line 50 of this file:

https://gist.github.com/Drevanoorschot/b5bc4298dc2f39bca5c5a17a3ea9d1c5#file-par_count-pvl-L50

I think it has to do with the expression being inside the indexation of the array. While using the first example solves the issue the second example poses, it may still be worth looking into.

**Comment by wytseoortwijn** [comment]
> I think it has to do with the expression being inside the indexation of the array.

No, this has to do with the use of non-linear arithmetic. As a rule of thumb, you have to consider yourself lucky if you are able to verify contracts with non-linear arithmetic.

Sometimes one can help the underlying solver a bit by rewriting non-linear subexpressions, which is exactly what you did. With the first line you simply were lucky, and with the second line you weren't.

**Comment by bobismijnnaam** [comment]
I tried the second formulation with a newer arithmetic solver (mentioned in #284 in this comment: https://github.com/utwente-fmt/vercors/pull/284#issuecomment-552858754) and then Vercors does verify it succesfully. It also verifies the first form succesfully. While this doesn't imply the newer arithmetic solver solved it because it can handle more cases (maybe the seed or order of solving changed or something) it might be nice to have a second alternative to try if you want to verify code like this. So if you run into this problem again I guess you can give the code in #284 a try.

**Comment by niomaster** [comment]
Yes, I like that idea. Different solvers/logics should remain available in z3 relatively stably, so I don't see a real issue. I think it would be especially nice if we could set different options per proof goal, but I think that would require quite a bit of effort on the viper side.

**Comment by bobismijnnaam** [comment]
>  but I think that would require quite a bit of effort on the viper side.

We could do something along the lines of splitverify and spin up multiple viper instances with multiple options, each assuming all methods it doesn't have to prove... But that sounds quite involved.

**Comment by niomaster** [comment]
Things we discussed that we might want to do, in order of priority:

* Make a note about non-linear arithmetic in the PVL syntax near operators that make sense
* Make a VerCors tutorial (#88) and include an explanation of this in the tutorial
* Investigate options of Z3 and/or add a flag to use the solver above
* Investigate whether it would be useful to set solver intrinsics per proof goal/method

**Comment by niomaster** [comment]
For now I'll mark the issue as doc, as that's what we assigned most priority.
